File: src/LambdaCalc/Expr.hs

module LambdaCalc.Expr where

type Name = String

data Expr
   = EVar Name
   | ELam Name Expr
   | EApp Expr Expr
   | EInt Int
   | EBool Bool
   | EUndefined
   | EPrim PrimOp Expr Expr
   | EFix Expr
   | ELet [(Name, Expr)] Expr
   | EIf Expr Expr Expr
   deriving(Show, Eq)

data PrimOp
   = Add
   | Sub
   | Mul
   deriving(Show, Eq)

File: cpp/src/Expr.cpp

#include "Expr.h"

using namespace lambdacalc;

Expr::~Expr()
{

}

EVar::EVar(const std::string &name) : _name(name)
{

}

void EVar::print(std::ostream &os) const
{
	os << "EVar " << _name;
}

ELam::ELam(const std::string &name, std::shared_ptr<Expr> expr) : _name(name), _expr(expr)
{

}

void ELam::print(std::ostream &os) const
{
	os << "ELam " << _name << " (" << *_expr << ")";
}

EApp::EApp(std::shared_ptr<Expr> left, std::shared_ptr<Expr> right) 
	: _left(left), _right(right)
{

}

void EApp::print(std::ostream &os) const
{
	os << "EApp (" << *_left << ") (" << *_right << ")";
}